### DISCUSSION TITLE
Accessing to a custom widget by using WidgetManager

### DISCUSSION THREAD

**Discussion by AhmetTanakol** [eclipse-theia/theia#9540] [Q&A]
Hello,

I am trying to access a custom widget by calling `getWidget(factoryId)` method of `WidgetManager`, however I don't see my widget id in `factories`. Is using widget manager correct way to access widgets? 

![screen shot 2018-08-06 at 11 50 54 am](https://user-images.githubusercontent.com/10375761/43710335-9f7ce3e4-996f-11e8-86dc-25315335c9d1.png)

**Comment by akosyakov** [comment]
Have you checked that a factory is registered and widgets are always created via `WidgetManager.getOrCreate`?

**Comment by AhmetTanakol** [comment]
I am calling `app.shell.addWidget` by giving an instance of my custom widget as the parameter. Do I need to use `WidgetManager`?

**Comment by akosyakov** [comment]
Calling `app.shell.addWidget` is fine. How do you create an instance of a widget? It is the responsibility of `WidgetManager`. If you create it not via `getOrCreate` then `WidgetManager` does not know about this widget.

**Comment by AhmetTanakol** [comment]
I see, then I should change the way I create the widget. But how can I pass some arguments that I can use them to set some fields in my widget's constructor.`getOrCreate ` takes only 2 parameters

**Comment by AhmetTanakol** [comment]
I assume I need to write setters in my widget class in order to set those fields

**Comment by akosyakov** [comment]
The second arg is to pass custom options.

**Comment by AhmetTanakol** [comment]
I am trying to pass a `URI` object in widget options however I get the following error

`widget-manager.ts:155 Uncaught (in promise) TypeError: Converting circular structure to JSON`

**Comment by akosyakov** [comment]
It should be JSON object. These options are stored as a part of the layout information and used to recover the widget on reload. A URI can be converted to the string and then converted back to a URI in the factory.

**Comment by AhmetTanakol** [comment]
Thank you, I managed to create what I need by extending `WidgetOpenHandler`

**Comment by akosyakov** [comment]
Right, it is the good base class if you want to provide a widget per a URI, like editors, previews and so on. Other notable abstractions:
- `NavigatableWidget` & `NavigatableWidgetOpenHandler` should be implemented if your widget is based on some resource, then it will be auto-revealed in the navigator, renamed properly on file moves and so on
- `SaveableWidget` should be implemented if your widget should support safe-lifecycle
- `AbstractViewContribution` - if you need a widget which should be a singleton per an application (view), e.g. navigator view, problems view and so on.

We should add code examples to our docs. For now, you are better to study files where these concepts are declared and their usages.
